src/Order/Checkout/CheckoutForm.php

```diff
@@ -0,0 +1,269 @@
+    ];
+
+    /**
+     * @var ServiceFactory
+     */
+    public $paymentServiceFactory;
+
+    /**
+     * @var SupportedCurrenciesInterface
+     */
+    public $supportedCurrencies;
+
+    /**
+     * @var Order
+     */
+    protected $cart;
+
+    /**
+     * @var null|string|false
+     */
+    protected $paymentError = null;
+
+    /**
+     * @inheritdoc
+     * @throws \Exception
+     */
+    public function __construct(Order $cart, ?RequestHandler $controller = null, ?string $name = null)
+    {
+        $this->cart = $cart;
+        $cart->Unlock();
+
+        parent::__construct(
+            $controller,
+            $name ?? static::DEFAULT_NAME,
+            $this->buildFields(),
+            $this->buildActions(),
+            CheckoutFormValidator::create()
+        );
+
+        if (!$this->getSessionData()) {
+            $this->extend('beforeLoadDataFromCart');
+            $this->loadDataFrom($cart);
+        }
+    }
+
+    /**
+     * @return Order
+     */
+    public function getCart(): Order
+    {
+        return $this->cart;
+    }
+
+    /**
+     * @return null|string
+     */
+    public function getPaymentError(): ?string
+    {
+        if ($this->paymentError !== null) {
+            // paymentError = false means cached result of "no error"
+            return $this->paymentError ?: null;
+        }
+
+        $paymentIdentifier = $this->getRequest()->getVar(static::PAYMENT_ID_QUERY_PARAM);
+
+        if (empty($paymentIdentifier)) {
+            // No payment query param
+            $this->paymentError = false;
+            return null;
+        }
+
+        /** @var Payment|null $payment */
+        $payment = $this->cart->Payments()->find('Identifier', $paymentIdentifier);
+        $defaultMessage = _t(self::class . '.PAYMENT_ERROR',
+            'There was an error processing your payment. Please try again.');
+
+        if ($payment === null) {
+            // No payment with that identifier for this order, can't show error
+            $errorMessage = false;
+        } elseif ($payment->Status === PaymentStatus::VOID) {
+            // Void status is returned for cancelled or declined card
+            $errorMessage = $defaultMessage;
+        } else {
+            /** @var PaymentMessage|null $message */
+            $message = $payment->Messages()->last();
+            $errorMessage = $message instanceof PurchaseError || $message instanceof CompletePurchaseError
+                ? $message->Message
+                : $defaultMessage;
+        }
+
+        $this->paymentError = $errorMessage;
+        return $errorMessage ?: null;
+    }
+
+    /**
+     * @param array $data
+     * @return HTTPResponse
+     * @throws \SilverStripe\Omnipay\Exception\InvalidConfigurationException
+     * @throws \SilverStripe\Omnipay\Exception\InvalidStateException
+     * @throws \Exception
+     */
+    public function ConfirmCheckout(array $data): HTTPResponse
+    {
+        $this->cart->Lock();
```

Comment:
We should be more consistent with using getters where they exist rather than direct property access.